django/utils/connection.py

```diff
@@ -0,0 +1,79 @@
+from copy import deepcopy
+
+from asgiref.local import Local
+
+from django.conf import settings as django_settings
+from django.utils.functional import cached_property
+
+
+class ConnectionProxy:
+    """Proxy for accessing a connection object's attributes."""
+
+    def __init__(self, connections, alias):
+        self.__dict__['_connections'] = connections
+        self.__dict__['_alias'] = alias
+
+    def __getattr__(self, item):
+        return getattr(self._connections[self._alias], item)
+
+    def __setattr__(self, name, value):
+        return setattr(self._connections[self._alias], name, value)
+
+    def __delattr__(self, name):
+        return delattr(self._connections[self._alias], name)
+
+    def __contains__(self, key):
+        return key in self._connections[self._alias]
+
+    def __eq__(self, other):
+        return self._connections[self._alias] == other
+
+
+class ConnectionDoesNotExist(Exception):
+    pass
+
+
+class BaseConnectionHandler:
+    settings_name = None
+    exception_class = ConnectionDoesNotExist
+    thread_critical = False
+
```

Comment:
Again, I suspect we could use `__slots__` here:

```suggestion
    __slots__ = ('_connections', '_settings')

```